src/System.Management.Automation/DscSupport/JsonDscClassCache.cs

```diff
@@ -0,0 +1,3249 @@
+        {
+            var className = cimClass.ClassName;
+            var alias = GetFriendlyName(cimClass);
+            var friendlyName = string.IsNullOrEmpty(alias) ? className : alias;
+            var moduleQualifiedResourceName = GetModuleQualifiedResourceName(module.Name, module.Version.ToString(), className, friendlyName);
+            ClassCache[moduleQualifiedResourceName].IsImportedImplicitly = false;
+        }
+
+        /// <summary>
+        /// Imports configuration keywords from a .psm1 file.
+        /// </summary>
+        /// <param name="module"></param>
+        /// <param name="resourceName"></param>
+        /// <param name="schemaFilePath"></param>
+        public static bool ImportScriptKeywordsFromModule(PSModuleInfo module, string resourceName, out string schemaFilePath)
+        {
+            return ImportScriptKeywordsFromModule(module, resourceName, out schemaFilePath, null);
+        }
+
+        /// <summary>
+        /// Imports configuration keywords from a .psm1 file.
+        /// </summary>
+        /// <param name="module"></param>
+        /// <param name="resourceName"></param>
+        /// <param name="schemaFilePath"></param>
+        /// <param name="functionsToDefine"></param>
+        public static bool ImportScriptKeywordsFromModule(PSModuleInfo module, string resourceName, out string schemaFilePath, Dictionary<string, ScriptBlock> functionsToDefine)
+        {
+            if (!ExperimentalFeature.IsEnabled(jsonSchemaSupportExperimentalFeatureName))
+            {
+                throw new InvalidOperationException(ParserStrings.PsDscJsonSchemaSupportDisabled);
+            }
+
+            if (module == null)
+            {
+                throw PSTraceSource.NewArgumentNullException(nameof(module));
+            }
+
+            if (resourceName == null)
+            {
+                throw PSTraceSource.NewArgumentNullException(nameof(resourceName));
+            }
+
+            schemaFilePath = Path.Join(module.ModuleBase, "DscResources", resourceName, resourceName + ".Schema.psm1");
+
+            if (File.Exists(schemaFilePath) && !s_currentImportingScriptFiles.Contains(schemaFilePath))
+            {
+                // If script dynamic keywords has already been loaded from the file, don't load them again.
+                // The ScriptKeywordFile cache is always initialized from scratch by the top-level
+                // configuration statement so within a single compile, things shouldn't change.
+                if (!ScriptKeywordFileCache.Contains(schemaFilePath))
+                {
+                    // Parsing the file is all that needs to be done to add the keywords
+                    // BUGBUG - need to fix up how the module gets set.
+                    // BUGBUG - should fail somehow if errors is not empty
+                    Token[] tokens; ParseError[] errors;
+                    s_currentImportingScriptFiles.Add(schemaFilePath);
```

Comment:
This hashset usage makes this public static API not threadsafe -- it might be better to use a `ConcurrentDictionary<string, bool>`

Reply:
This was copy-paste from existing mof-code, which does not seem to have problems with this.